File: RCallerPhp/ROutputParser.php
<?php

/*
 *
  RCallerPhp, A solution for calling R from Php
  Copyright (C) 2010,2011  Mehmet Hakan Satman

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 *
 * Mehmet Hakan Satman - mhsatman@yahoo.com
 * http://www.mhsatman.com
 * Google code projec: http://code.google.com/p/rcaller/
 *
 */

/**
 *
 * @author Mehmet Hakan Satman
 */
class ROutputParser {

  private $XMLFile = null;
  private $XMLContent = null;

  public function getXMLFile() {
    return $this->XMLFile;
  }

  public function getXMLFileAsString() {
    return file_get_contents($this->XMLFile);
    ;
  }

  public function setXMLFile($XMLFile) {
    $this->XMLFile = $XMLFile;
  }

  public function getXMLContent() {
    return ($this->XMLContent);
  }

  public function parse() {
    $this->XMLContent = simplexml_load_file($this->getXMLFile());
  }

  function __construct() {
    
  }

  public function getNames() {
    $names = null;
    $i = 0;
    foreach ($this->XMLContent->variable as $v) {
      $names[$i] = (String) $v->attributes()->name;
      $i++;
    }
    return($names);
  }

  public function getAsStringArray($name) {
    $values = null;
    foreach ($this->XMLContent->variable as $v) {
      if ((String) $v->attributes()->name == $name) {
        for ($i = 0; $i < count($v->value); $i++) {
          $values[$i] = (String) $v->value[$i];
        }
      }
    }
    return($values);
  }

  /*
    public double[][] getAsDoubleMatrix(String name, int n, int m) throws RCallerParseException {
    double[][] result = new double[n][m];
    double[] arr = this.getAsDoubleArray(name);
    int c = 0;
    for (int i = 0; i < n; i++) {
    for (int j = 0; j < m; j++) {
    result[i][j] = arr[c];
    c++;
    }
    }
    return (result);
    }
   */
}

?>

File: RCaller/src/main/java/com/github/rcaller/rstuff/ROutputParser.java
/*
 *
RCaller, A solution for calling R from Java
Copyright (C) 2010-2014  Mehmet Hakan Satman

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 *
 * Mehmet Hakan Satman - mhsatman@yahoo.com
 * http://www.mhsatman.com
 * Google code project: https://github.com/jbytecode/rcaller
 * Please visit the blog page with rcaller label:
 * http://stdioe.blogspot.com.tr/search/label/rcaller
 */
package com.github.rcaller.rstuff;

import com.github.rcaller.exception.ExecutionException;
import com.github.rcaller.exception.ParseException;
import com.github.rcaller.exception.XMLParseException;
import com.github.rcaller.io.ArrowBridge;
import com.github.rcaller.io.ROutputParserArrow;
import com.github.rcaller.io.ROutputParserXML;
import org.w3c.dom.Document;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.InputSource;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import java.io.*;
import java.net.URI;
import java.util.ArrayList;

/**
 * API for reading R IPC output.
 * @author Mehmet Hakan Satman
 */
public interface ROutputParser {

    /**
     * Static factory implementation creator with given options.
     * Returned implementation uses XML or Arrow IPC format.
     * @param rCallerOptions given options
     * @return ROutputParser implementing object
     */
    static ROutputParser create(RCallerOptions rCallerOptions) {
        if (rCallerOptions.useArrowIfAvailable()) {
            if (ArrowBridge.isArrowAvailable(rCallerOptions)) {
                //Use Arrow by default if enabled
                return new ROutputParserArrow(rCallerOptions);
            } else {
                if (rCallerOptions.failIfArrowNotAvailable()) {
                    throw new ExecutionException("Arrow is enabled but not available");
                } else {
                    //Use XML if Arrow is enabled but not available and we should not fail
                    return new ROutputParserXML();
                }
            }
        } else {
            //Use XML if Arrow is disabled
            return new ROutputParserXML();
        }
    }

    /**
     * Returns parsed XML document.
     * @deprecated Do not use if you are not using XML consciously
     * @return parsed XML document
     * @throw NotImplementedException if Arrow implementation is used
     */
    @Deprecated
    Document getDocument();

    /**
     * Inits parsed XML document.
     * @deprecated Do not use if you are not using XML consciously
     * @throw NotImplementedException if Arrow implementation is used
     */
    @Deprecated
    void setDocument(Document document);

    /**
     * Returns XML file.
     * @deprecated Do not use if you are not using XML consciously. Can be replaced with {@link #getIPCResource}
     * @return XML file
     * @throw NotImplementedException if Arrow implementation is used
     */
    @Deprecated
    File getXMLFile();

    URI getIPCResource();

    /**
     * Returns raw XML document.
     * @deprecated Do not use if you are not using XML consciously
     * @return raw XML document
     * @throw NotImplementedException if Arrow implementation is used
     */
    @Deprecated
    String getXMLFileAsString() throws IOException;

    /**
     * Inits XML file.
     * @deprecated Do not use if you are not using XML consciously. Can be replaced with {@link #setIPCResource}
     * @throw NotImplementedException if Arrow implementation is used
     */
    @Deprecated
    void setXMLFile(File xmlFile);

    void setIPCResource(URI ipcResourceURI);

    void parse() throws ParseException;

    ArrayList<String> getNames();

    String getType(String variablename);

    int[] getDimensions(String name);

    /**
     * Returns nodes from parsed XML document.
     * @deprecated Do not use if you are not using XML consciously
     * @return nodes from parsed XML document
     * @throw NotImplementedException if Arrow implementation is used
     */
    @Deprecated
    NodeList getValueNodes(String name);

    String[] getAsStringArray(String name) throws ParseException;

    double[] getAsDoubleArray(String name) throws ParseException;

    float[] getAsFloatArray(String name) throws ParseException;

    int[] getAsIntArray(String name) throws ParseException;

    long[] getAsLongArray(String name) throws ParseException;

    default boolean[] getAsLogicalArray(String name) throws ParseException {
        String[] strResults = getAsStringArray(name);
        boolean[] bools = new boolean[strResults.length];
        for (int i = 0; i < strResults.length; i++) {
            try {
                bools[i] = Boolean.parseBoolean(strResults[i]);
            } catch (Exception e) {
                throw new ParseException("String value '" + strResults[i] + "' can not convert to boolean");
            }
        }
        return (bools);
    }

    double[][] getAsDoubleMatrix(String name, int n, int m) throws ParseException;

    double[][] getAsDoubleMatrix(String name) throws ParseException;

}
